### PR TITLE
feat: wave/7 gap fixes — all 9 Shepherd audit items

### PR DESCRIPTION
**Pull request by dean0x**
## Summary

- **P0: Zero-stderr fix** — Redirect `check_hook_version_mismatch()` from `eprintln!` to `hook_log::log_hook_warning()`, preserving the zero-stderr invariant in hook mode
- **P0: HookProtocol dispatch** — Wire `protocol_for_agent()` factory so all 6 agents (Claude, Cursor, Gemini, Copilot, Codex, OpenCode) use their protocol's `parse_input()`/`format_response()` instead of hardcoded Claude JSON
- **P1: SHA-256 integrity** — `skim agents` now uses `verify_script_integrity()` for hook status (ok/tampered/missing/unknown)
- **P1: Hook timeout** — 5-second watchdog thread calls `process::exit(0)` to prevent agent hangs
- **P1: Collision detection** — Init warns when other Bash PreToolUse hooks exist (informational, not blocking)
- **P1: Multi-agent init** — `skim init --agent <name>` with per-agent config dir resolution, script generation, and agent-aware uninstall
- **P2: Session dedup** — `dedup_invocations()` by (command_key, timestamp) across agents
- **P2: Awareness tracking** — SHA-256 hash manifest for awareness files, `--force` required if user modified
- **P2: Agent-not-found** — Helpful error with install hints when target agent isn't installed

## Test plan

- [ ] `cargo test --all-features` — all pass (28 new tests)
- [ ] `cargo clippy -- -D warnings` — zero warnings
- [ ] `cargo fmt -- --check` — clean
- [ ] Manual: `echo '{"tool_input":{"command":"cargo test"}}' | skim rewrite --hook` → Claude JSON, empty stderr
- [ ] Manual: `echo '{"tool_input":{"command":"cargo test"}}' | skim rewrite --hook --agent gemini` → Gemini JSON, empty stderr
- [ ] Manual: `echo '{"command":"cargo test"}' | skim rewrite --hook --agent cursor` → Cursor JSON, empty stderr
- [ ] Manual: `skim init --agent codex --yes` → awareness-only message
- [ ] Manual: `skim init --yes --dry-run` → Claude backward compat
- [ ] Manual: `skim agents --json` → integrity field uses SHA-256

### CODE REVIEW

File: crates/rskim/src/cmd/hooks/cursor.rs [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+    }
+
+    fn parse_input(&self, json: &serde_json::Value) -> Option<HookInput> {
+        // Cursor puts command at top level, not nested under tool_input
+        let command = json.get("command").and_then(|c| c.as_str())?.to_string();
+        Some(HookInput { command })
+    }
+
+    fn format_response(&self, rewritten_command: &str) -> serde_json::Value {
+        serde_json::json!({
```
Comment (dean0x) [inline crates/rskim/src/cmd/hooks/cursor.rs]: **[SECURITY] Cursor hook sets permission field breaking security invariant (HIGH: 92% confidence)**

The project security invariant (documented in `cmd/init/mod.rs`) states: "The hook NEVER sets \`permissionDecision\`. Unlike competitors, our hook only sets \`updatedInput\` and lets Claude Code's permission system evaluate independently."

The Cursor hook's `format_response` sets `"permission": "allow"`. While Cursor may use `"permission"` instead of `"permissionDecision"`, this is a semantic deviation that auto-approves the rewritten command instead of letting Cursor's permission system decide.

**Fix**: Document whether Cursor requires the `"permission"` field for `updated_input` to take effect, or if it can be omitted (letting Cursor's own permission system decide). If required, add an explicit comment:
```rust
// NOTE: Cursor's permission system interprets permission: allow to bypass
// user confirmation. This is intentional — we trust rewritten commands automatically.
// If Cursor changes its semantics, this may need revision.
fn format_response(&self, rewritten_command: &str) -> serde_json::Value {
```

If the field is NOT required, remove it entirely to maintain the documented security invariant.

File: crates/rskim/src/cmd/learn.rs [hunk 2]
```diff
@@ -640,24 +684,54 @@ fn sanitize_command_for_rules(cmd: &str) -> String {
         .to_string()
```
Comment (dean0x) [inline crates/rskim/src/cmd/learn.rs]: **[REGRESSION] Rules file filename change breaks existing automation (HIGH: 90% confidence)**

`skim learn --generate` output changed from `.claude/rules/cli-corrections.md` to `.claude/rules/skim-corrections.md`. Any existing scripts, CI pipelines, or users referencing the old path will silently create a new file instead of updating the existing one, leaving the old file orphaned but still active (duplicate rules).

**Fix**: Add migration logic in `write_rules_file()`:
```rust
let legacy_path = rules_dir.join("cli-corrections.md");
if legacy_path.exists() && !rules_path.exists() {
    std::fs::rename(&legacy_path, &rules_path)?;
    println!("Migrated: cli-corrections.md -> {}", filename);
}
```

This ensures users upgrading to the new version don't lose their existing corrections file.

**Comment by dean0x** [comment]
## PR #94 Review Summary: Consolidated Findings

**Review Scope**: 8 security/architecture/performance/complexity/consistency/regression/tests/rust reports covering wave/7 changes (41 files, 10,482 LOC added)

**Inline Comments Created**: 7 high-confidence findings (80%+ confidence)
- Inline comments posted for specific lines requiring code changes
- See individual comments above for detailed fixes

**Summary Statistics**

| Aspect | Blocking | Should Fix | Suggestions | Score |
|--------|----------|-----------|-------------|-------|
| Security | 2 HIGH | 2 MEDIUM | 3 LOW | 7/10 |
| Architecture | 3 HIGH | 2 MEDIUM | 3 LOW | 7/10 |
| Performance | 1 HIGH | 2 MEDIUM | 1 LOW | 7/10 |
| Complexity | 2 HIGH | 2 MEDIUM | 3 LOW | 6/10 |
| Consistency | 2 HIGH | 3 MEDIUM | 3 LOW | 7/10 |
| Regression | 2 HIGH | 1 MEDIUM | 2 LOW | 8/10 |
| Tests | 2 HIGH | 2 MEDIUM | 3 LOW | 8/10 |
| Rust | 2 HIGH | 3 MEDIUM | 3 LOW | 7/10 |

**Deduplication Notes**: Many issues appeared in multiple review reports. Below are consolidations:

### Duplicated Across Reviewers

**process::exit(0) safety** (Confidence: 90%)
- Flagged by: Security, Rust reviews
- Locations: rewrite.rs:1047
- Status: Inline comment posted
- Recommendation: Add SAFETY comment documenting the tradeoff

**Double read of settings.json** (Confidence: 90%)
- Flagged by: Performance review
- Locations: init/state.rs lines 40 & 63
- Status: Inline comment posted
- Recommendation: Pass parsed JSON to avoid re-read/re-parse

**Duplicated rate-limiting pattern** (Confidence: 88%)
- Flagged by: Complexity, Consistency reviews
- Locations: rewrite.rs lines 1195-1222 & 1246-1267
- Status: Not yet inline (moderate priority)
- Recommendation: Extract rate_limited_log_warning() helper

**Unbounded recursion in dir_size_bytes** (Confidence: 80%)
- Flagged by: Security, Performance reviews
- Locations: agents.rs lines 567-618
- Status: Not yet inline (moderate priority)
- Recommendation: Add max_depth parameter (10 levels)

**Repeated dirs::home_dir() calls** (Confidence: 82%)
- Flagged by: Performance review
- Locations: agents.rs lines 108, 146, 192, 236
- Status: Not yet inline
- Recommendation: Compute once, pass through

**Duplicated architecture patterns** (Confidence: 90%)
- Flagged by: Architecture review
- Issues:
  1. Hook detection logic in agents.rs vs init/state.rs
  2. Config dir resolution in rewrite.rs vs init/helpers.rs
  3. parse_input JSON paths across Claude/Gemini/Copilot
- Status: Not inline (requires refactoring, not trivial fixes)
- Recommendation: Consolidate into shared modules before adding more agents

### Lower-Confidence Findings (60-79%) Not Posted as Inline

These are worth investigating but fall below the inline-comment threshold:

**Architecture Issues**
- `AgentKind` living in session/types.rs creates awkward dependencies (65% confidence)
- `HookProtocol` trait forces dead-code install/uninstall stubs (85% confidence) — should split into separate traits
- Open/Closed Principle violation in agent dispatch (match statements in 3 places) (70% confidence)

**Performance Optimizations**
- Redundant split + join of command string (70% confidence)
- Box<dyn HookProtocol> unnecessary for zero-sized unit structs (82% confidence)
- session/mod.rs: get_providers(Some(kind)) probes all 6 agents (80% confidence)

**Consistency Issues**
- Hook install/uninstall comment variations (85% confidence) — should standardize message
- Missing doc comments on 4 of 6 hook structs (82% confidence)
- Inconsistent symlink warning behavior across providers (83% confidence)

**Rust Idiom**
- `AgentKind::from_str` shadows standard FromStr trait (85% confidence) — should implement trait
- `#[allow(dead_code)]` annotations excessive in volume (80% confidence) — mask real dead code
- `resolve_agent_name()` hardcoded to "claude-code" (82% confidence) — should use agent_kind parameter

**Test Issues**
- Environment variable mutation in test not thread-safe (92% confidence) — use subprocess isolation
- Log file assertion in tamper test is conditional (85% confidence) — should assert unconditionally
- Redundant zero-stderr tests (82% confidence) — remove 4 individual tests, keep parameterized

**Regression Risks**
- Version field in suggest output changed from "1.0.0" to dynamic (85% confidence, likely intentional bug fix)
- Hook version mismatch moved from stderr to log (intentional improvement per GRANITE #361)

### Pre-existing Issues (Not Blocking)

- Verify_script_integrity returns Ok(true) for missing manifest (backward compat design)
- rewrite.rs is 2,608 lines (pre-wave/7 bloat)
- AgentKind location in session/types.rs creates awkward dependencies
- Missing integration test for watchdog timeout behavior

---

## Summary Assessment

**Code Quality**: Solid. Well-structured multi-agent architecture with clear trait-based dispatch patterns. Excellent separation of concerns in the init module refactor.

**Blocking Issues**: 3 findings require code changes before merge:
1. process::exit SAFETY comment (rewrite.rs:1047)
2. Cursor hook "permission" field documentation (hooks/cursor.rs:29)
3. Filename migration for learn --generate (learn.rs:684)

**Should-Fix Items**: 8 additional issues worth addressing in this wave if time permits:
1. Double read of settings.json (init/state.rs)
2. Cursor hook security variance documentation
3. Duplicated rate-limiting pattern (rewrite.rs)
4. Claude hook test naming inconsistency
5. Codex session provider 6-level nesting
6. OpenCode stub test naming (use_noop → use_stub)
7. Unbounded recursion in dir_size_bytes (agents.rs)
8. Environment variable mutation in test (agents.rs)

**Architecture Debt**: The multi-agent dispatch pattern is repeated in 3 modules (agents/hooks/session). Consider a registration-based pattern before adding more agents.

**Testing**: 1,522 tests (up from 1,495). Good integration coverage for new features. Missing: watchdog timeout behavioral test, mixed-agent dedup test.

---

**PR Status**: Ready for discussion. Recommend addressing the 3 blocking findings + high-priority should-fix items before merge. The lower-confidence findings can be tracked as tech debt for Phase 3.

---

_Review conducted by Claude Code (AI assistant). This summary consolidates findings from 8 review reports. Inline comments are scoped to >=80% confidence findings._

File: crates/rskim/src/cmd/init/install.rs [hunk 3]
```diff
@@ -0,0 +1,545 @@
...
+                );
+            } else {
+                println!("  {} Updated: {}", check_mark(true), script_path.display());
+            }
+        }
+    } else {
+        println!("  {} Created: {}", check_mark(true), script_path.display());
+    }
+
+    // Generate script content
```
Comment (dean0x) [inline crates/rskim/src/cmd/init/install.rs]: **Architecture: generate_script() dead code -- install.rs bypasses trait** (85% confidence)

`create_hook_script()` formats the hook script content inline with `format!()` (lines 331-337). Meanwhile, each `HookProtocol` implementation has a `generate_script()` method that produces the same agent-specific content. The install path never calls `protocol.generate_script()`.

This creates two sources of truth for hook script content. The trait method is effectively dead code in production -- only exercised in unit tests.

**Fix**: Wire `create_hook_script()` to call `hooks::protocol_for_agent(agent).generate_script(binary_path, version)` for the script body.

File: crates/rskim/src/cmd/init/install.rs [hunk 4]
```diff
@@ -0,0 +1,545 @@
+//! Install flow for `skim init`.
```
Comment (dean0x) [inline crates/rskim/src/cmd/init/install.rs]: ## Wave/7 Code Review Summary

Reviewed 10 reports across 42 changed files (+10,768 / -1,244 lines). This PR introduces multi-agent support (6 agents) with session providers, hook protocols, integrity verification, and shell redirect support.

### Overall Assessment

✅ **APPROVED WITH CONDITIONS** - All findings are non-blocking (no CRITICAL issues). The architecture is sound, test coverage is comprehensive, and the refactoring of the monolithic 1,047-line init.rs into focused sub-modules is exemplary.

### Review Findings Consolidated

**Blocking Issues (≥80% confidence):**
- **17 inline comments created** covering high-confidence findings across security, architecture, complexity, performance, consistency, and regression categories
- All findings are addressed with concrete, actionable fixes

**High-Confidence Issue Counts:**
| Category | Issues | Confidence |
|----------|--------|-----------|
| Security | 3 | 82-85% |
| Architecture | 4 | 85-92% |
| Performance | 2 | 83-85% |
| Complexity | 3 | 85-92% |
| Consistency | 2 | 90-92% |
| Rust | 3 | 80-85% |
| Tests | 1 | 90% |
| Regression | 2 | 82-85% |

### Key Strengths

1. ✅ **Zero unsafe code** - Entire diff uses safe Rust exclusively
2. ✅ **Error handling** - All production code uses Result/? pattern, no unwrap() in business logic
3. ✅ **Security mindset** - Multiple layers of defense (size guards, symlink checks, bounded I/O, rate limiting)
4. ✅ **Test coverage** - 400+ tests passing with excellent behavior-focused integration tests
5. ✅ **Module decomposition** - init.rs refactor from 1,047 lines into 5 focused modules is exemplary
6. ✅ **Strategy Pattern** - HookProtocol and SessionProvider traits provide clean multi-agent abstraction
7. ✅ **Performance** - Meets <50ms target (14.6ms for 3000-line files, 3x faster than target)

### Lower-Confidence Issues (60-79%)

The following findings from review reports are below the 80% threshold for inline comments but should be noted:

- **Hash manifest written without restricted permissions** (68%) - consider chmod 0600 on .sha256 files
- **Copilot YAML deserialization without limits** (62%) - serde_yaml_ng may lack DoS protections
- **Agent CLI name unsanitized in path construction** (65%) - add debug_assert on path separators
- **Levenshtein watchdog thread spawns on every invocation** (70%) - consider lazy spawn after stdin read
- **Multiple test edge cases** (65-70%) - provider neutralization duplication, test naming precision, rotation testing gaps

### Backward Compatibility

✅ **All major regressions verified as safe:**
- `skim init` defaults to ClaudeCode (SAFE)
- `skim rewrite --hook` defaults to ClaudeCode (SAFE)
- `skim learn` migration logic included (cli-corrections.md → skim-corrections.md)
- Session dedup is additive (SAFE for single-agent deployments)

**Note**: Three behavioral changes require release notes documentation:
1. Hook warnings now route to ~/.cache/skim/hook.log (was stderr)
2. skim_hook_version now reports actual version (was hardcoded "1.0.0")
3. Rules filename changed to skim-corrections.md

### Recommendation

**Merge with condition**: Address the 17 inline comments (concrete fixes provided for each). These are mechanical improvements that strengthen code quality without architectural changes. None block merge.

---

*Review: Security (8/10), Architecture (7/10), Rust (8/10), Performance (8/10), Complexity (6/10), Consistency (8/10), Tests (8/10), Regression (8/10)*

*Generated by Claude Code review agent*
